### PR TITLE
Use score instead of likelihood in Vanilla Gradients

### PR DESCRIPTION
**Pull request by AlexKubiesa**
Addresses #159 

I've modified VanillaGradients.explain so instead of using the full model, it first creates a "score model" - the same as the original model, except without the final activation layer - and uses that instead.

If callers supply a model where the last layer isn't an activation layer, a warning is generated, and we use the original model as the score model. This tends to produce uninteresting output, so it should be pretty clear to users that something has gone wrong in this case.

Calling VanillaGradients.explain directly from VanillaGradientsCallback caused tf.function rerouting because the score model was being regenerated on every epoch. This made Tensorflow generate lots of warning messages and presumably reduces performance (I think I did see a performance hit when I tried it locally). So, instead, I split VanillaGradientsCallback.explain into "get_score_model" and "explain_score_model". The idea is that get_score_model gets called only once by the callback, and explain_score_model gets called every epoch. This is possible because the base Callback class has a set_model method which gets called by the framework, and I added an override for it.

- [x] Added code is tested
- [x] I've run Black and linter

### CODE REVIEW

File: tf_explain/core/vanilla_gradients.py [hunk 1]
```diff
@@ -1,13 +1,29 @@
...
+followed by an activation layer."
+
+_activation_layer_classes = (
+    tf.keras.layers.Activation,
+    tf.keras.layers.LeakyReLU,
+    tf.keras.layers.PReLU,
+    tf.keras.layers.ReLU,
+    tf.keras.layers.Softmax,
+    tf.keras.layers.ThresholdedReLU,
+)
```
Comment (RaphaelMeudec) [inline tf_explain/core/vanilla_gradients.py:24]: Maybe you can check dynamically that the layer is in `tf.keras.activations`. So we don't have to store all the activation types on our side

File: tests/conftest.py [hunk 2]
```diff
@@ -55,3 +55,52 @@ def convolutional_model(random_data):
     model.compile(optimizer="adam", loss="categorical_crossentropy")
 
     return model
+
+
+@pytest.fixture()
+def convolutional_model_for_vanilla_gradients(random_data):
```
Comment (RaphaelMeudec) [inline tests/conftest.py:61]: I think naming could be better with something that really explains what's the difference for vanilla gradients. Something like `convolutional_model_with_independant_activation_layer`

File: tests/conftest.py [hunk 3]
```diff
@@ -55,3 +55,52 @@ def convolutional_model(random_data):
...
+        ]
+    )
+
+    model.compile(optimizer="adam", loss="categorical_crossentropy")
+
+    return model
+
+
+@pytest.fixture()
+def score_model_for_vanilla_gradients(random_data):
```
Comment (RaphaelMeudec) [inline tests/conftest.py:86]: Same, maybe `convolutional_model_without_final_activation`

File: tf_explain/core/vanilla_gradients.py [hunk 4]
```diff
@@ -24,15 +40,66 @@ def explain(self, validation_data, model, class_index):
         Args:
             validation_data (Tuple[np.ndarray, Optional[np.ndarray]]): Validation data
                 to perform the method on. Tuple containing (x, y).
-            model (tf.keras.Model): tf.keras model to inspect
+            model (tf.keras.Model): tf.keras model to inspect. The last two layers of
+                the model should be: a layer which computes class scores with no
+                activation, followed by an activation layer.
```
Comment (RaphaelMeudec) [inline tf_explain/core/vanilla_gradients.py:45]: I think it would be nice to explain why we have to do that

File: tf_explain/core/vanilla_gradients.py [hunk 5]
```diff
@@ -56,19 +123,13 @@ def compute_gradients(images, model, class_index):
...
-
         with tf.GradientTape() as tape:
             inputs = tf.cast(images, tf.float32)
             tape.watch(inputs)
-            predictions = model(inputs)
-            loss = tf.keras.losses.categorical_crossentropy(
-                expected_output, predictions
-            )
+            scores = model(inputs)
+            scores_for_class = scores[:, class_index : class_index + 1]
```
Comment (RaphaelMeudec) [inline tf_explain/core/vanilla_gradients.py:130]: Why `scores[:, class_index : class_index + 1]` and not just `scores[:, class_index]` ?

File: tf_explain/core/vanilla_gradients.py [hunk 6]
```diff
@@ -1,13 +1,29 @@
...
 
 from tf_explain.utils.display import grid_display
 from tf_explain.utils.image import transform_to_normalized_grayscale
 from tf_explain.utils.saver import save_grayscale
 
 
+UNSUPPORTED_ARCHITECTURE_WARNING = "\
+Unsupported model architecture for VanillaGradients. The last two layers of \
+the model should be: a layer which computes class scores with no activation, \
+followed by an activation layer."
```
Comment (RaphaelMeudec) [inline tf_explain/core/vanilla_gradients.py:15]: It might be a personal opinion, but I'm not a fan of the line breakers. Could you format it this way instead?

```python
UNSUPPORTED_ARCHITECTURE_WARNING = (
    "Unsupported model architecture for VanillaGradients. The last two layers of "
    "the model should be: a layer which computes class scores with no activation, "
    "followed by an activation layer."
)
```

File: tests/core/test_vanilla_gradients.py [hunk 7]
```diff
@@ -1,20 +1,47 @@
...
+    softmax_layer = convolutional_model_for_vanilla_gradients.layers[-1]
+    score_model = explainer.get_score_model(convolutional_model_for_vanilla_gradients)
+    # The score model should exclude the final activation layer.
+    assert score_model.layers[-1] == score_layer
+    assert softmax_layer not in score_model.layers
+
+
+def test_get_score_model_logs_warnings_when_model_not_suitable(convolutional_model):
+    explainer = VanillaGradients()
+    with pytest.warns(UserWarning, match=r"^Unsupported model architecture"):
```
Comment (RaphaelMeudec) [inline tests/core/test_vanilla_gradients.py:46]: Maybe you can import `UNSUPPORTED_ARCHITECTURE_WARNING` to make sure there is an exact match? It avoids test to fail if we change the formulation in the future.

**Comment by RaphaelMeudec** [comment]
Thanks for the PR @AlexKubiesa! I think the way you've done it is nice, I just put some minor comments on the PR. Thanks again!

**Comment by RaphaelMeudec** [comment]
@AlexKubiesa could you rebase your PR on master? I've fixed the failing tests, and migrated from Travis to Github Actions. Thanks!

File: tf_explain/core/vanilla_gradients.py (hunk 1 above)
Comment (AlexKubiesa) [reply tf_explain/core/vanilla_gradients.py:24]: I don't think that would work because the things in `tf.keras.activations` are activation functions rather than layers. Unfortunately there isn't really anything to distinguish an activation layer in tf.keras from any other type of layer...

But I assume the vast majority of models people will be applying Vanilla Gradients to will have softmax as the last activation function so as long as we have `Activation` and `Softmax` on this list, I think we will catch most cases.

File: tests/conftest.py (hunk 3 above)
Comment (AlexKubiesa) [reply tests/conftest.py:86]: Yes, good point! I'll rename these.

**Comment by AlexKubiesa** [comment]
Thanks for having a look @RaphaelMeudec, I'll push again once I've addressed all the comments.

By the way, it looks like the tox.ini file is gone from master, which means tox doesn't run for me locally after I rebased. Is that intended?

File: tf_explain/core/vanilla_gradients.py (hunk 4 above)
Comment (AlexKubiesa) [reply tf_explain/core/vanilla_gradients.py:45]: Ok, I'll add an explanation.

File: tf_explain/core/vanilla_gradients.py (hunk 6 above)
Comment (AlexKubiesa) [reply tf_explain/core/vanilla_gradients.py:15]: Ah, I didn't know you could do it that way!

File: tests/core/test_vanilla_gradients.py (hunk 7 above)
Comment (AlexKubiesa) [reply tests/core/test_vanilla_gradients.py:46]: Ok

**Comment by AlexKubiesa** [comment]
I've rebased now and I think I've addressed all the comments.

**Comment by RaphaelMeudec** [comment]
All good! Thanks for the PR @AlexKubiesa!

**Review by RaphaelMeudec — APPROVED** [review]
